src/rgw/rgw_rados.cc

```diff
@@ -6913,15 +6957,19 @@ int RGWRados::block_while_resharding(RGWRados::BucketShard *bs,
   constexpr int num_retries = 10;
   for (int i = 1; i <= num_retries; i++) { // nb: 1-based for loop
     auto& ref = bs->bucket_obj.get_ref();
-    ret = cls_rgw_get_bucket_resharding(ref.pool.ioctx(), ref.obj.oid, &entry);
+    if (ref.obj.empty()) {
+      ret = -ENOENT;
```

Comment:
That is the case where the old bucket index object was deleted prior to the call to BucketShard::init on line 6892. That is, 
we get ENOENT at https://github.com/ceph/ceph/pull/52097/files#diff-29d082098be66d6c3c4ee097c51077f4021620ceeae88f74bd719b9e5b4d1eadR6892 and bs->bucket_obj is not initialized to an actual object before we arrive here.

Reply:
ok. `BucketShard::init()` shouldn't be sending any operation on the index shard itself, it would only try to read the bucket instance metadata. so this handles the case where the old bucket instance was deleted, and we rely on `fetch_new_bucket_id()` below to load the new one based on the bucket entrypoint metadata

Reply:
yes exactly

Reply:
thanks. that makes sense, but it's subtle enough to warrant a comment